### PR TITLE
Add custom implementation of covariance

### LINKED ISSUE #4: Implement Covariance with DP (without combine)

### PR DESCRIPTION
**Pull request by LaRiffle**
## Description

Fix #4 

Implem according to our discussion on Notion

@Jasopaum how do we generate the doc automatically again? I realize I didn't do it for the variance...

## Definition of Done

- [x] I followed the [Arkhn Code Book](https://www.notion.so/arkhn/Arkhn-Code-Book-23ac18a8af7343d0a3f61f1c9ef7400c) (I swear!).
- [ ] I have written tests for the code I added or updated.
- [x] I have updated the documentation according to my changes.
- [x] I have updated the deployment configuration if needed.

### CODE REVIEW

File: inst/python/custom_dp/covariance_custom.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def noisy_average(x, epsilon, x_min, x_max, N, N_noisy):
+    """
+    Implement the NoisyAverage with Normalization algorithm
+    from https://books.google.fr/books?id=WFttDQAAQBAJ&pg=PA29.
+
+    We assume that N and N_noisy are already computed
+    """
+    assert x_min <= x_max, "Verify that you bounds are corrected ordered"
```
Comment (Jasopaum) [inline inst/python/custom_dp/covariance_custom.py:16]: Even if I'm not sure how reticulate uses Python code, it's usually better to use `if _: raise(_)` than `assert` outside of tests, one reason being that these `asserts` are removed when the compilation is optimized

File: inst/python/custom_dp/covariance_custom.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import numpy as np
+
+
+def lap(scale):
+    """Compute Laplacian noise"""
+    return np.random.laplace(0, scale)
+
+
+def noisy_average(x, epsilon, x_min, x_max, N, N_noisy):
```
Comment (Jasopaum) [inline inst/python/custom_dp/covariance_custom.py:9]: We usually try avoid variable names starting with an upper-case character

File: inst/python/custom_dp/covariance_custom.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    xy_max = max(x_min_ * y_min_, x_min_ * y_max_, x_max_ * y_min_, x_max_ * y_max_)
+
+    # Shift the variables and compute their product
+    xy_ = (x - E_x) * (y - E_y)
+
+    # Compute the noisy average of the normalized product, ie the covariance
+    E_xy = noisy_average(xy_, eps, xy_min, xy_max, N, N_noisy)
+
+    # N - 1 is used to have an unbiased estimate (like standard implem in numpy)
+    E_xy = N_noisy / (N_noisy - 1) * E_xy
```
Comment (Jasopaum) [inline inst/python/custom_dp/covariance_custom.py:63]: Can you return directly or use another name to store this?

Comment (Jasopaum) [inline inst/python/custom_dp/covariance_custom.py:63]: Also, `E_xy * N_noisy / (N_noisy - 1)` is less confusing, imo

**Review by Jasopaum — COMMENTED** [review]
I'm not sure where the $- E[X]E[Y]$ from $Cov(X, Y) = E[XY] - E[X]E[Y]$ is implemented 🤔

File: inst/python/custom_dp/covariance_custom.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    assert x_min <= x_max, "Verify that you bounds are corrected ordered"
+
+    # Crop the data using the bounds provided
+    x = np.clip(x, x_min, x_max)
+
+    normalized_sum_x = sum(x) - N * (x_min + x_max) / 2 + lap((x_max - x_min) / epsilon)
+
+    if N_noisy <= 1:
+        E_x = (x_min + x_max) / 2
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply inst/python/custom_dp/covariance_custom.py:24]: Good point!

**Comment by Jasopaum** [comment]
> Actually what I do is first replacing Unable to render expression. x and Unable to render expression. y with their centered version using this line:
> 
> ```
> xy_ = (x - E_x) * (y - E_y)
> ```
> 
> So technically I don't compute `E[XY]` but `E[(X - E(X))*(Y - E(Y)]` which is equal to `E[XY]−E[X]E[Y]`.
> 
> Does it look good to you?

I forgot how to derive `E[(X - E(X))*(Y - E(Y)] = E[XY]−E[X]E[Y]` but if it's the case, I'm fine with that 😅

**Pull request merged**